Frequency analysis

Mafusn opened the pull request:
This pull request introduces a frequency analysis pipeline for homophonic substitution ciphers, including core analysis logic, feasibility checks, backtracking search, and comprehensive tests. It also refines constants for improved precision, updates example cipher data, and adjusts file loading for better path handling.

**Core Analysis and Algorithm Implementation:**

* Added `find_letter_candidates` in `src/analyzer/analyze.py` to perform frequency-based candidate symbol set search for each letter, integrating feasibility checks and backtracking.
* Implemented `backtracking` in `src/analyzer/backtracking.py` to efficiently find subsets of cipher symbols whose frequencies match target letter frequencies, with pruning and candidate ranking.
* Added feasibility utilities in `src/analyzer/feasibility.py` to compute target frequency ranges and check combinatorial feasibility before running expensive searches.

**Testing and Validation:**

* Introduced unit and performance tests for `find_letter_candidates` (`tests/analyzer/test_analyze.py`), `backtracking` (`tests/analyzer/test_backtracking.py`), and feasibility logic (`tests/analyzer/test_feasibility.py`) to ensure correctness and efficiency. [[1]](diffhunk://#diff-8ef3044d72c30ad942e7a30a8779ac6ecdfa5970cf9ef74b4f0689954b392212R1-R55) [[2]](diffhunk://#diff-1851bb6c79d109a07acde7cdd184cee44aaa63d890519f8e4757861575b71885R1-R44) [[3]](diffhunk://#diff-eb904b07b5a7052a4f23a793ae16494a407873d50db2fa0cdbfde81a5c52acccR1-R32)
* Updated integration tests for cipher loading to use new file path conventions and the updated example cipher file. [[1]](diffhunk://#diff-92129da1d3c95ade67f8dcf811ad4fc206dcec9908e96b7bdb20d40d4a80f244L1-L11) [[2]](diffhunk://#diff-92129da1d3c95ade67f8dcf811ad4fc206dcec9908e96b7bdb20d40d4a80f244L38-R39)

**Data and Constants Updates:**

* Added new example cipher files with plaintext, keys, and ciphertext in `data/example_ciphers/`. [[1]](diffhunk://#diff-b6b414959395bac89153d6d76c06fd22bce79d407769bee552b529c26c44da39R1-R176) [[2]](diffhunk://#diff-460691066d7333389c5af9e9d07399d680c03bc9900996f888296e2030966a62R1-R92)
* Refined frequency and range tolerances in `src/constants.py` for more precise candidate selection, and ensured all code uses updated constants and path handling. [[1]](diffhunk://#diff-42d0f00c4780305311eae71fecca993a0523587cb4156e039cabd0d1ad8dbbb8L4-R5) [[2]](diffhunk://#diff-e27ebfc826ce6bc8a7a212ef5a9b56bfabf18ddb025219bea233049de949dafcR3-R8) [[3]](diffhunk://#diff-e27ebfc826ce6bc8a7a212ef5a9b56bfabf18ddb025219bea233049de949dafcL19-R21) [[4]](diffhunk://#diff-51246e53255db77c9edad496f074aa1bdbf8dbdc11f89a02040115c9ab4fa7f0L3-R3)

**Miscellaneous and Documentation:**

* Updated the `README.md` to reflect new commands and testing coverage options.
* Modified `src/main.py` to use the new analysis pipeline as the entry point.

These changes collectively establish a robust, test-driven foundation for analyzing and attacking homophonic ciphers using frequency analysis and combinatorial search.

Mafusn committed:
Feat: load_cipher() takes type Path instead of str

Mafusn committed:
Fix: Runs the project as a module

Mafusn committed:
Feat: Homophonic analyzer

Mafusn committed:
Update analyze.py

Mafusn committed:
Fix: loading of cipher is updated to match json format

Mafusn committed:
Fix: Can go through 1 trillion combinations

Mafusn committed:
Feat: finds more candidates

Mafusn committed:
Feat: Keeps only top 10 candidates

Mafusn committed:
Test: tests for homophonic analyzer

Mafusn committed:
Test: Perfomance test for analyzer

Mafusn committed:
Feat: candidates are returned as a list of tuples

List[Tuple[str, List[Set[int]]]]

Fantaskink commented on README.md:
Use backticks  for commands.
Example: 
- \`uv run python -m src.main`
- `uv run python -m src.main`

Mafusn committed:
Fix: Now we have backticks

Fantaskink reviewed (approved):
LGTM mr fuss

Jatewo committed:
fix: Update test command and pytest configuration

Simplified the test command in README.md to use the default pytest invocation. Updated pyproject.toml to specify test paths and add pytest options for verbose output and coverage reporting.

Jatewo committed:
feat: Print letter candidates in main function

Updated the main function to print each letter and its candidate sets returned by find_letter_candidates, providing more detailed output for analysis.

Jatewo reviewed (approved):
LGTM dawg

Fantaskink reviewed (approved):
LGTM^2
